fish-rust/src/curses.rs

```diff
@@ -0,0 +1,261 @@
+                // so we can safely unwrap here.
+                result => CStr::from_ptr(result),
+            }
+        };
+        Some(ArcCow::Borrowed(result))
+    }
+}
+
+impl<'a> Capability<'a> for StringCap {
+    type Result = Option<ArcCow<'a, CStr, CString>>;
+
+    fn lookup(&self, term: &'a Term) -> Self::Result {
+        if term.has_overrides.load(Ordering::Relaxed) {
+            let overrides = term.overrides.lock().expect("Mutex poisoned!");
+            if let Ok(idx) = overrides.binary_search_by(|entry| entry.0.cmp(self)) {
+                return Some(ArcCow::Shared(Arc::clone(&overrides[idx].1)));
+            }
+        }
+
+        self.sys_lookup()
+    }
+}
+
+impl<'a> Capability<'a> for Number {
+    type Result = Option<i32>;
+
+    fn lookup(&self, _: &'a Term) -> Self::Result {
+        unsafe {
+            match tgetnum(self.0.as_ptr()) {
+                -1 => None,
+                n => Some(n),
+            }
+        }
+    }
+}
+
+impl<'a> Capability<'a> for Flag {
+    type Result = bool;
+
+    fn lookup(&self, _: &'a Term) -> Self::Result {
+        unsafe { tgetflag(self.0.as_ptr()) }
+    }
+}
+
+/// Calls the curses `setupterm()` function with the provided `$TERM` value `term` (or a null
+/// pointer in case `term` is null) for the file descriptor `fd`.
+///
+/// Note that the `errret` parameter is provided to the function, meaning curses will not write
+/// error output to stderr in case of failure.
+///
+/// Any existing references from `curses::term()` will be invalidated by this call!
+pub fn setup(term: Option<&str>, fd: i32) -> bool {
+    let result = unsafe {
+        // If cur_term is already initialized for a different $TERM value, calling setupterm() again
+        // will leak memory. Call reset() first to free previously allocated resources.
+        reset();
+
+        let mut err = 0;
+        if let Some(term) = term {
+            let term = CString::new(term).unwrap();
+            sys::setupterm(term.as_ptr(), fd, &mut err)
+        } else {
+            sys::setupterm(core::ptr::null(), fd, &mut err)
+        }
+    };
+
+    unsafe {
+        if result == sys::OK {
+            TERM = Lazy::new(Term::new);
+            true
+        } else {
+            TERM = Lazy::new(|| panic!("TERM has not yet been initialized!"));
+            false
+        }
+    }
+}
+
+/// Whether or not the curses library has been initialized.
+pub fn is_initialized() -> bool {
+    unsafe { !sys::cur_term.is_null() }
+}
+
+/// Resets the curses `cur_term` TERMINAL pointer. Any previous term objects are invalidated!
+pub unsafe fn reset() {
+    if is_initialized() {
+        sys::del_curterm(cur_term);
+        sys::cur_term = core::ptr::null();
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct Code {
+    /// The two-char termcap code for the capability, followed by a nul.
+    code: [u8; 3],
+}
+
+impl Code {
```

Comment:
That's not untrue. I personally thought the ergonomic wins and more importantly, the maintainability with no one ever forgetting the trailing `\0` or accidentally using the long-form terminfo name instead of the short-form termcap name was worth the extra dozen SLOC.